server/analyzer/optimize_functions.go

```diff
@@ -31,12 +31,35 @@ func OptimizeFunctions(ctx *sql.Context, a *analyzer.Analyzer, node sql.Node, sc
 	if scope != nil && scope.CurrentNodeIsFromSubqueryExpression {
 		return node, transform.SameTree, nil
 	}
-	return pgtransform.NodeExprsWithNodeWithOpaque(node, func(node sql.Node, expr sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
+	hasSRF := false
+	funcName := ""
+	var function sql.Expression
+	var tempType sql.Type
+	node, same, err := pgtransform.NodeExprsWithNodeWithOpaque(node, func(node sql.Node, expr sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
 		if compiledFunction, ok := expr.(*framework.CompiledFunction); ok {
+			funcName = compiledFunction.Name
+			hasSRF = compiledFunction.IsSRF()
+			if hasSRF {
+				// TODO: need a way to get the return type from the function
+				//  this is temporary way to get the return type
+				tempType = compiledFunction.Arguments[0].Type()
```

Comment:
This only works for the functions that are included in testing since the set type's internal type always matches the first argument's type, but this is guaranteed to fail for any other function. The return type of the function should contain everything we need.